From 0ff778875e15e42bedfde9d33dea6674d4da7325 Mon Sep 17 00:00:00 2001
From: William Loh <wloh@google.com>
Date: Tue, 30 Aug 2022 00:11:15 +0000
Subject: [PATCH] Limit length and number of MIME types you can set

Limit character length of MIME types to 255. If this length is exceeded
then a IllegalArugmentException is thrown. The number of MIME types that
can be set is also limited to 500 per MIME group with the number of
total MIME Groups also limited to 500. A IllegalStateException is thrown if this number is exceeded.

Bug: 237291548
Test: Installed and ran POC app from b/237291548
Change-Id: I1d57e674f778cfacdc89225ac3273c432a39af63
Merged-In: I1d57e674f778cfacdc89225ac3273c432a39af63
(cherry picked from commit 3ae3406b9706163073c282a8c4081faa32b606b2)
Merged-In: I1d57e674f778cfacdc89225ac3273c432a39af63
---
 .../android/content/pm/parsing/ParsingPackageImpl.java   | 3 +++
 .../core/java/com/android/server/pm/PackageSetting.java  | 9 +++++++++
 2 files changed, 12 insertions(+)

diff --git a/core/java/android/content/pm/parsing/ParsingPackageImpl.java b/core/java/android/content/pm/parsing/ParsingPackageImpl.java
index 5a7f21040d0a..a6e1867fce0c 100644
--- a/core/java/android/content/pm/parsing/ParsingPackageImpl.java
+++ b/core/java/android/content/pm/parsing/ParsingPackageImpl.java
@@ -1695,6 +1695,9 @@ public class ParsingPackageImpl implements ParsingPackage, Parcelable {
         for (int i = component.getIntents().size() - 1; i >= 0; i--) {
             IntentFilter filter = component.getIntents().get(i);
             for (int groupIndex = filter.countMimeGroups() - 1; groupIndex >= 0; groupIndex--) {
+                if (mimeGroups != null && mimeGroups.size() > 500) {
+                    throw new IllegalStateException("Max limit on number of MIME Groups reached");
+                }
                 mimeGroups = ArrayUtils.add(mimeGroups, filter.getMimeGroup(groupIndex));
             }
         }
diff --git a/services/core/java/com/android/server/pm/PackageSetting.java b/services/core/java/com/android/server/pm/PackageSetting.java
index 376326264d9c..d8f902a9a086 100644
--- a/services/core/java/com/android/server/pm/PackageSetting.java
+++ b/services/core/java/com/android/server/pm/PackageSetting.java
@@ -328,11 +328,20 @@ public class PackageSetting extends PackageSettingBase {
     }
 
     public boolean setMimeGroup(String mimeGroup, List<String> mimeTypes) {
+        for (String mimeType : mimeTypes) {
+            if (mimeType.length() > 255) {
+                throw new IllegalArgumentException("MIME type length exceeds 255 characters");
+            }
+        }
         ArraySet<String> oldMimeTypes = getMimeGroupInternal(mimeGroup);
         if (oldMimeTypes == null) {
             throw new IllegalArgumentException("Unknown MIME group " + mimeGroup
                     + " for package " + name);
         }
+        if (mimeTypes.size() > 500) {
+            throw new IllegalStateException("Max limit on MIME types for MIME group "
+                    + mimeGroup + " exceeded for package " + name);
+        }
 
         ArraySet<String> newMimeTypes = new ArraySet<>(mimeTypes);
         boolean hasChanges = !newMimeTypes.equals(oldMimeTypes);
-- 
2.39.0.rc1.256.g54fd8350bd-goog

